<div align="center">
    <img src="https://images.unsplash.com/photo-1551288049-bebda4e38f71?ixlib=rb-4.0.3&auto=format&fit=crop&w=1200&q=80" alt="E-commerce Analytics" width="100%" style="border-radius: 10px;">
</div>

<br>

# 🇧🇷 Olist Brazilian E-Commerce Analytics

Welcome to this interactive Exploratory Data Analysis (EDA) notebook! 
Olist is a Brazilian e-commerce platform. In this project, we will clean a massive dataset of +100k orders and use **Plotly** to uncover business insights regarding sales trends, customer geography, and peak shopping hours.

### 🎯 Project Goals:
* **Data Unification:** Merge multiple relational datasets into a single source of truth.
* **Data Cleaning:** Perform rigorous preprocessing (Handling Nulls, dropping irrelevant columns, and standardizing data).
* **Business Insights:** Extract actionable recommendations using Interactive Visualizations.

---

---
## Import & read data

In [1]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.io as pio
pio.renderers.default = "iframe_connected"
pio.templates.default = "plotly_dark"
import warnings
warnings.filterwarnings('ignore')

In [2]:
customers = pd.read_csv("/kaggle/input/datasets/organizations/olistbr/brazilian-ecommerce/olist_customers_dataset.csv")
geolocation = pd.read_csv("/kaggle/input/datasets/organizations/olistbr/brazilian-ecommerce/olist_geolocation_dataset.csv")
items = pd.read_csv("/kaggle/input/datasets/organizations/olistbr/brazilian-ecommerce/olist_order_items_dataset.csv")
payments = pd.read_csv("/kaggle/input/datasets/organizations/olistbr/brazilian-ecommerce/olist_order_payments_dataset.csv")
reviews = pd.read_csv("/kaggle/input/datasets/organizations/olistbr/brazilian-ecommerce/olist_order_reviews_dataset.csv")
orders_dataset = pd.read_csv("/kaggle/input/datasets/organizations/olistbr/brazilian-ecommerce/olist_orders_dataset.csv")
products = pd.read_csv("/kaggle/input/datasets/organizations/olistbr/brazilian-ecommerce/olist_products_dataset.csv")
sellers = pd.read_csv("/kaggle/input/datasets/organizations/olistbr/brazilian-ecommerce/olist_sellers_dataset.csv")
translation = pd.read_csv("/kaggle/input/datasets/organizations/olistbr/brazilian-ecommerce/product_category_name_translation.csv")

In [3]:
customers.head(3)

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP


In [4]:
geolocation.head(3)

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP


In [5]:
items.head(3)

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.9,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.9,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.0,17.87


In [6]:
payments.head(3)

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71


In [7]:
reviews.head(3)

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24


In [8]:
orders_dataset.head(3)

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00


In [9]:
products.head(3)

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0


In [10]:
sellers.head(3)

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ


In [11]:
translation.head(3)

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto


## 🔗 Merging Datasets
### The Olist dataset is built like a relational database. To get a comprehensive view of our business, we need to join these tables into a single, unified DataFrame. 
### We will merge the orders table with customers, items, products (translated to English), sellers, payments, and reviews using their respective primary keys.

In [12]:
payments_summary = payments.groupby('order_id', as_index=False)['payment_value'].sum()

reviews_summary = reviews.groupby('order_id', as_index=False)['review_score'].mean()

products_en = pd.merge(products, translation, on='product_category_name', how='left')

In [13]:
df = pd.merge(orders_dataset, customers, on='customer_id', how='left')

df = pd.merge(df, items, on='order_id', how='left')

df = pd.merge(df, products_en, on='product_id', how='left')

df = pd.merge(df, sellers, on='seller_id', how='left')

df = pd.merge(df, payments_summary, on='order_id', how='left')

df = pd.merge(df, reviews_summary, on='order_id', how='left')
df

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,customer_unique_id,customer_zip_code_prefix,...,product_weight_g,product_length_cm,product_height_cm,product_width_cm,product_category_name_english,seller_zip_code_prefix,seller_city,seller_state,payment_value,review_score
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00,7c396fd4830fd04220f754e42b4e5bff,3149,...,500.0,19.0,8.0,13.0,housewares,9350.0,maua,SP,38.71,4.0
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00,af07308b275d755c9edb36a90c618231,47813,...,400.0,19.0,13.0,19.0,perfumery,31570.0,belo horizonte,SP,141.46,4.0
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00,3a653a41f6f9fc3d2a113cf8398680e8,75265,...,420.0,24.0,19.0,21.0,auto,14840.0,guariba,SP,179.12,5.0
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00,7c142cf63193a1473d2e66489a9ae977,59296,...,450.0,30.0,10.0,20.0,pet_shop,31842.0,belo horizonte,MG,72.20,5.0
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00,72632f0f9dd73dfee390c9b22eb56dd6,9195,...,250.0,51.0,15.0,15.0,stationery,8752.0,mogi das cruzes,SP,28.62,5.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
113420,63943bddc261676b46f01ca7ac2f7bd8,1fca14ff2861355f6e5f14306ff977a7,delivered,2018-02-06 12:58:58,2018-02-06 13:10:37,2018-02-07 23:22:42,2018-02-28 17:37:56,2018-03-02 00:00:00,da62f9e57a76d978d02ab5362c509660,11722,...,4950.0,40.0,10.0,40.0,baby,17602.0,tupa,SP,195.00,4.0
113421,83c1379a015df1e13d02aae0204711ab,1aa71eb042121263aafbe80c1b562c9c,delivered,2017-08-27 14:46:43,2017-08-27 15:04:16,2017-08-28 20:52:26,2017-09-21 11:24:17,2017-09-27 00:00:00,737520a9aad80b3fbbdad19b66b37b30,45920,...,13300.0,32.0,90.0,22.0,home_appliances_2,8290.0,sao paulo,SP,271.01,5.0
113422,11c177c8e97725db2631073c19f07b62,b331b74b18dc79bcdf6532d51e1637c1,delivered,2018-01-08 21:28:27,2018-01-08 21:36:21,2018-01-12 15:35:03,2018-01-25 23:32:54,2018-02-15 00:00:00,5097a5312c8b157bb7be58ae360ef43c,28685,...,6550.0,20.0,20.0,20.0,computers_accessories,37175.0,ilicinea,MG,441.16,2.0
113423,11c177c8e97725db2631073c19f07b62,b331b74b18dc79bcdf6532d51e1637c1,delivered,2018-01-08 21:28:27,2018-01-08 21:36:21,2018-01-12 15:35:03,2018-01-25 23:32:54,2018-02-15 00:00:00,5097a5312c8b157bb7be58ae360ef43c,28685,...,6550.0,20.0,20.0,20.0,computers_accessories,37175.0,ilicinea,MG,441.16,2.0


In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113425 entries, 0 to 113424
Data columns (total 32 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   order_id                       113425 non-null  object 
 1   customer_id                    113425 non-null  object 
 2   order_status                   113425 non-null  object 
 3   order_purchase_timestamp       113425 non-null  object 
 4   order_approved_at              113264 non-null  object 
 5   order_delivered_carrier_date   111457 non-null  object 
 6   order_delivered_customer_date  110196 non-null  object 
 7   order_estimated_delivery_date  113425 non-null  object 
 8   customer_unique_id             113425 non-null  object 
 9   customer_zip_code_prefix       113425 non-null  int64  
 10  customer_city                  113425 non-null  object 
 11  customer_state                 113425 non-null  object 
 12  order_item_id                 

---
## 🧹 1. Data Cleaning & Preprocessing
Before analyzing the data, we must ensure its quality. In this section we will:
* Filter only completed/delivered orders.
* Handle missing values intelligently without losing crucial data.
* Translate and standardize Brazilian State abbreviations to full names for better mapping.

In [15]:
date_columns = [
    'order_purchase_timestamp', 
    'order_approved_at', 
    'order_delivered_carrier_date', 
    'order_delivered_customer_date', 
    'order_estimated_delivery_date',
    'shipping_limit_date'
]

for col in date_columns:
    df[col] = pd.to_datetime(df[col])

print(df[date_columns].dtypes)

order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    datetime64[ns]
order_estimated_delivery_date    datetime64[ns]
shipping_limit_date              datetime64[ns]
dtype: object


In [16]:
df.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 161
order_delivered_carrier_date     1968
order_delivered_customer_date    3229
order_estimated_delivery_date       0
customer_unique_id                  0
customer_zip_code_prefix            0
customer_city                       0
customer_state                      0
order_item_id                     775
product_id                        775
seller_id                         775
shipping_limit_date               775
price                             775
freight_value                     775
product_category_name            2378
product_name_lenght              2378
product_description_lenght       2378
product_photos_qty               2378
product_weight_g                  793
product_length_cm                 793
product_height_cm                 793
product_width_cm                  793
product_cate

In [17]:
df['order_status'].value_counts()

order_status
delivered      110197
shipped          1186
canceled          706
unavailable       610
invoiced          361
processing        357
created             5
approved            3
Name: count, dtype: int64

In [18]:
px.pie(df, names='order_status', title='Distribution of Order Statuses').show()

In [19]:
df = df.dropna(subset=['product_id'])
df['product_category_name_english'] = df['product_category_name_english'].fillna('Unknown')
df['review_score'] = df['review_score'].fillna(0)

In [20]:
df.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                  15
order_delivered_carrier_date     1194
order_delivered_customer_date    2454
order_estimated_delivery_date       0
customer_unique_id                  0
customer_zip_code_prefix            0
customer_city                       0
customer_state                      0
order_item_id                       0
product_id                          0
seller_id                           0
shipping_limit_date                 0
price                               0
freight_value                       0
product_category_name            1603
product_name_lenght              1603
product_description_lenght       1603
product_photos_qty               1603
product_weight_g                   18
product_length_cm                  18
product_height_cm                  18
product_width_cm                   18
product_cate

## 🛠️ Handling Missing Values (The Business Logic)

### Data cleaning is not just about removing empty rows; it is about distinguishing between **data errors** and **business facts**. We adopted a strategic approach to preserve the context of our operations:

### 1. **The "Natural" Nulls (Operational Insights):** We kept the missing values in `order_delivered_customer_date` (2,454 rows) and `order_delivered_carrier_date` (1,194 rows). These are **not errors**. They represent orders that were either `canceled` or `unavailable` and therefore never reached the delivery stage. Dropping these would bias our analysis and hide critical operational bottlenecks.

### 2. **Minor Administrative Gaps:** The 15 missing values in `order_approved_at` are negligible and do not impact our macro-level analysis, so they were left as is to maintain data integrity.

### 3. **Critical Data Filtering:** We strictly dropped rows missing essential features for financial and physical calculations (e.g., `payment_value` or `product_weight_g`), ensuring that our revenue and logistics insights remain accurate.

### 4. **Categorical Fill:** Missing product categories were labeled as `'Unknown'` rather than dropping the entire record. This allows us to keep the transaction in our sales volume reports while acknowledging the missing taxonomy.

In [21]:
cols_to_drop = ['product_weight_g', 'payment_value']
df = df.dropna(subset=cols_to_drop)

cols_to_zero = ['product_name_lenght', 'product_description_lenght', 'product_photos_qty']
for col in cols_to_zero:
    df[col] = df[col].fillna(0)
    
df['product_category_name'] = df['product_category_name'].fillna('Unknown')

In [22]:
df.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                  15
order_delivered_carrier_date     1194
order_delivered_customer_date    2454
order_estimated_delivery_date       0
customer_unique_id                  0
customer_zip_code_prefix            0
customer_city                       0
customer_state                      0
order_item_id                       0
product_id                          0
seller_id                           0
shipping_limit_date                 0
price                               0
freight_value                       0
product_category_name               0
product_name_lenght                 0
product_description_lenght          0
product_photos_qty                  0
product_weight_g                    0
product_length_cm                   0
product_height_cm                   0
product_width_cm                    0
product_cate

In [23]:
df['seller_state'].unique()

array(['SP', 'MG', 'ES', 'RS', 'DF', 'PR', 'SC', 'RJ', 'GO', 'BA', 'MA',
       'AC', 'PB', 'PE', 'CE', 'MT', 'PI', 'RN', 'MS', 'PA', 'AM', 'SE',
       'RO'], dtype=object)

In [24]:
brazil_states = {
    'SP': 'São Paulo',
    'MG': 'Minas Gerais',
    'ES': 'Espírito Santo',
    'RS': 'Rio Grande do Sul',
    'DF': 'Distrito Federal',
    'PR': 'Paraná',
    'SC': 'Santa Catarina',
    'RJ': 'Rio de Janeiro',
    'GO': 'Goiás',
    'BA': 'Bahia',
    'MA': 'Maranhão',
    'PB': 'Paraíba',
    'PE': 'Pernambuco',
    'CE': 'Ceará',
    'MT': 'Mato Grosso',
    'PI': 'Piauí',
    'RN': 'Rio Grande do Norte',
    'MS': 'Mato Grosso do Sul',
    'PA': 'Pará',
    'AM': 'Amazonas',
    'SE': 'Sergipe',
    'RO': 'Rondônia'
}

df['seller_state'] = df['seller_state'].replace(brazil_states)

df['customer_state'] = df['customer_state'].replace(brazil_states)

df['seller_state'].unique()

df['customer_state'].unique()

array(['São Paulo', 'Bahia', 'Goiás', 'Rio Grande do Norte', 'Paraná',
       'Rio Grande do Sul', 'Rio de Janeiro', 'Minas Gerais',
       'Santa Catarina', 'RR', 'Pernambuco', 'TO', 'Ceará',
       'Distrito Federal', 'Sergipe', 'Mato Grosso', 'Paraíba', 'Pará',
       'Rondônia', 'Espírito Santo', 'AP', 'Mato Grosso do Sul',
       'Maranhão', 'Piauí', 'AL', 'AC', 'Amazonas'], dtype=object)

In [25]:
missing_states = {
    'RR': 'Roraima',
    'TO': 'Tocantins',
    'AP': 'Amapá',
    'AL': 'Alagoas',
    'AC': 'Acre'
}

df['customer_state'] = df['customer_state'].replace(missing_states)

df['customer_state'].unique()

array(['São Paulo', 'Bahia', 'Goiás', 'Rio Grande do Norte', 'Paraná',
       'Rio Grande do Sul', 'Rio de Janeiro', 'Minas Gerais',
       'Santa Catarina', 'Roraima', 'Pernambuco', 'Tocantins', 'Ceará',
       'Distrito Federal', 'Sergipe', 'Mato Grosso', 'Paraíba', 'Pará',
       'Rondônia', 'Espírito Santo', 'Amapá', 'Mato Grosso do Sul',
       'Maranhão', 'Piauí', 'Alagoas', 'Acre', 'Amazonas'], dtype=object)

# EDA

---
## 📊 2. Exploratory Data Analysis (EDA)

### 📈 2.1 Sales Trend Over Time
Let's look at the big picture. How have order volumes changed since the company started?

In [26]:
filtered_df = df[(df['order_purchase_timestamp'] >= '2017-01-01') & 
                 (df['order_purchase_timestamp'] <= '2018-08-31')]

filtered_df['year_month'] = filtered_df['order_purchase_timestamp'].dt.to_period('M').astype(str)
monthly_orders = filtered_df.groupby('year_month').size().reset_index(name='order_count')
monthly_orders = monthly_orders.sort_values('year_month')

fig = px.line(
    monthly_orders, 
    x='year_month', 
    y='order_count', 
    title='📉 Monthly Orders Trend (Jan 2017 - Aug 2018)',
    markers=True,
    labels={'year_month': 'Month', 'order_count': 'Total Orders'}
)

fig.update_traces(line_color='#1f77b4', line_width=3)
fig.show()

**📌 Insight:**
*We can observe a continuous growth in orders over time, with significant peaks likely related to holiday seasons or massive discount events like Black Friday.*

In [27]:
top_cities = df['customer_city'].value_counts().head(10).reset_index()
top_cities.columns = ['customer_city', 'order_count']

fig = px.bar(
    top_cities, 
    x='customer_city', 
    y='order_count',
    title='🏙️ Top 10 Cities by Orders',
    text='order_count',
    color='order_count',
    color_continuous_scale='Blues'
)

fig.update_traces(textposition='outside')
fig.show()

**📌 Insight:**
*Unsurprisingly, **São Paulo** completely dominates the customer base, followed by Rio de Janeiro. This geographic concentration means that establishing distribution centers or optimizing supply chains in the Southeast region of Brazil will yield the highest return on investment (ROI) and significantly reduce delivery times.*

In [28]:
delivered_orders = df[df['order_status'] == 'delivered']
top_categories_revenue = delivered_orders.groupby('product_category_name_english')['price'].sum().sort_values(ascending=False).head(10).reset_index()

top_categories_revenue = df.groupby('product_category_name_english')['price'].sum().sort_values(ascending=False).head(10).reset_index()

fig = px.bar(
    top_categories_revenue, 
    x='price',
    y='product_category_name_english', 
    orientation='h',
    title='🏆 Top 10 Categories by Revenue',
    text_auto='.2s',
    color='price',
    color_continuous_scale='Viridis'
)

fig.update_layout(yaxis={'categoryorder':'total ascending'})
fig.show()

**📌 Insight:**
*While some items might sell in high quantities, the **Health & Beauty** and **Watches/Gifts** categories are the true cash cows, generating the most revenue. Marketing budgets should heavily target these high-margin categories.*

In [29]:
failed_orders = df[df['order_status'].isin(['canceled', 'unavailable'])]

top_failed_categories = failed_orders.groupby('product_category_name_english').size().reset_index(name='failed_count')
top_failed_categories = top_failed_categories.sort_values('failed_count', ascending=False).head(10)

fig_failed = px.bar(
    top_failed_categories,
    x='failed_count',
    y='product_category_name_english',
    orientation='h',
    title='🚨 Top 10 Categories by Failed Orders (Canceled & Unavailable)',
    text='failed_count',
    color='failed_count',
    color_continuous_scale='Reds'
)

fig_failed.update_layout(yaxis={'categoryorder':'total ascending'})
fig_failed.show()

**📌 Insight:**
*By keeping the non-delivered orders, we discovered critical operational bottlenecks. The categories shown in red above suffer from the highest cancellation and unavailability rates. The business needs to immediately investigate the inventory and supply chain partners handling these specific categories to reduce revenue leakage.*

In [30]:
review_dist = df['review_score'].value_counts().reset_index()
review_dist.columns = ['review_score', 'count']

review_dist['review_score'] = review_dist['review_score'].astype(str) + ' Stars'

fig = px.pie(
    review_dist, 
    values='count', 
    names='review_score',
    title='⭐ Review Score Distribution',
    hole=0.4, 
    color_discrete_sequence=px.colors.qualitative.Pastel
)

fig.update_traces(textposition='inside', textinfo='percent+label')
fig.show()

**📌 Insight:**
*The majority of customers are highly satisfied, with nearly **60% giving a 5-star rating**. However, there is a noticeable chunk (around 11%) of 1-star reviews. As a next step for the business, investigating these 1-star reviews (e.g., correlating them with late delivery times) would be crucial to improve customer retention.*

---
### ⏰ 2.4 Peak Shopping Hours (Heatmap)
When do Brazilians prefer to shop online? By extracting the day and hour from the timestamp, we can identify the busiest times for the platform.

In [31]:
df['order_hour'] = df['order_purchase_timestamp'].dt.hour
df['order_day'] = df['order_purchase_timestamp'].dt.day_name()

days_order = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']

heatmap_data = df.groupby(['order_day', 'order_hour']).size().reset_index(name='orders_count')

heatmap_matrix = heatmap_data.pivot(index='order_day', columns='order_hour', values='orders_count')
heatmap_matrix = heatmap_matrix.reindex(days_order)

fig = px.imshow(
    heatmap_matrix,
    labels=dict(x="Hour", y="Day", color="Number of orders"),
    x=heatmap_matrix.columns,
    y=heatmap_matrix.index,
    title='🔥 Heatmap of Peak Order Times',
    color_continuous_scale='YlOrRd'
)

fig.show()

**📌 Insight:**
*The heatmap clearly shows that peak shopping activity occurs during **weekdays (Monday to Friday) between 10:00 AM and 4:00 PM**. This suggests that many customers browse and buy during their work hours or lunch breaks! The marketing team should schedule their heavy ad campaigns during these exact hours.*

---
## 💡 Final Conclusions & Recommendations
1. **Sales are Growing:** The platform is experiencing healthy growth, but logistics must scale accordingly.
2. **Geographic Concentration:** A massive chunk of customers lives in São Paulo. Expanding warehouses near this state will dramatically reduce shipping costs and times.
3. **Marketing Timing:** Ads and flash sales should be targeted on weekdays during the mid-day (10 AM - 4 PM) for maximum conversion rates.

**Thank you for reading! If you found this notebook helpful, please consider giving it an Upvote! 👍**